lcls-twincat-motion/Library/POUs/Motion/FB_MotionRequest.TcPOU

```diff
@@ -0,0 +1,154 @@
+	With just the DUT_MotionStage/FB_MotionStage setup it is possible for two function blocks to
+	fight with and interfere with each other and with the EPICS commands.
+*)
+VAR_IN_OUT
+	// Motor to move
+    stMotionStage: DUT_MotionStage;
+END_VAR
+VAR_INPUT
+	// Start move on rising edge, stop move on falling edge
+    bExecute: BOOL;
+	// Reset errors on rising edge
+    bReset: BOOL;
+	// Define behavior for when the motor is already moving
+    enumMotionRequest: ENUM_MotionRequest := ENUM_MotionRequest.WAIT;
+	// Goal position
+    fPos: LREAL;
+	// Move velocity
+    fVel: LREAL;
+	// Optional acceleration
+    fAcc: LREAL;
+	// Optional deceleration
+    fDec: LREAL;
+END_VAR
+VAR_OUTPUT
+	// True if in error state
+    bError: BOOL;
+	// Error code, usually
+    sErrorMessage: STRING;
+	// If TRUE, we are moving the motor
+    bBusy: BOOL;
+	// If TRUE, we are not moving the motor and our most recent move was successful
+    bDone: BOOL;
+END_VAR
+VAR
+    rtExec: R_TRIG;
+    ftExec: F_TRIG;
+    rtReset: R_TRIG;
+    ftBusy: F_TRIG;
+    nState: UINT := 0;
+    nQueuedState: UINT;
+    bMyMove: BOOL;
+END_VAR
+// Define local constants for our state machine states
+VAR CONSTANT
+    INIT: UINT := 0;
+    EXEC: UINT := 1;
+    START: UINT := 2;
+    MOVING: UINT := 3;
+    DONE: UINT := 4;
+    STOP: UINT := 5;
+    ERROR: UINT := 6;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[rtExec(CLK:=bExecute);
+ftExec(CLK:=bExecute);
+rtReset(CLK:=bReset);
+
+// Go back to INIT state on reset
+IF rtReset.Q THEN
+    nState := INIT;
+    nQueuedState := INIT;
+END_IF
+IF rtExec.Q OR ftExec.Q THEN
+	bDone := FALSE;
+END_IF
+
+CASE nState OF
+	// Start by setting everything to a known value
+    INIT:
+        nState := EXEC;
+        bError := FALSE;
+        sErrorMessage := '';
+        bBusy := FALSE;
+        bDone := FALSE;
+	// Normal "waiting for move" state
+    EXEC:
+        bMyMove := FALSE;
+		// Looking for a rising edge on bExecute
+        IF rtExec.Q THEN
+			bDone := FALSE;
+            IF stMotionStage.bExecute THEN
+				// Pick interruption behavior via state machine if a move is in progress
+                CASE enumMotionRequest OF
+                    ENUM_MotionRequest.WAIT:
+                        nState := MOVING;
+                        nQueuedState := START;
+                    ENUM_MotionRequest.INTERRUPT:
+                        nState := STOP;
+                        nQueuedState := START;
+                    ENUM_MotionRequest.ABORT:
+                        nState := EXEC;
+                END_CASE
+            ELSE
+                nState := START;
+            END_IF
+        END_IF
+	// Set the correct values on DUT_MotionStage to start a new absolute move
+    START:
+        bMyMove := TRUE;
+        stMotionStage.bExecute := TRUE;
+        stMotionStage.nCommand := ENUM_EpicsMotorCmd.MOVE_ABSOLUTE;
+        stMotionStage.fPosition := fPos;
```

Comment:
I like this a lot. In https://github.com/jsheppard95/lcls-twincat-optics/commit/9af138bd91bb6ca87189853490b61182d81602bf I've been trying to emulate this by writing the the `DUT_MotionStage` attributes after a move requests and then pass the instance to `FB_MotionStage` but its easy to get the timing messed up. It looks like it will be much easier using this as an interface. I will try and see if this leads me to new comments.